### PR TITLE
Fix proxy usage and creation in IceGrid implementation

### PR DESCRIPTION
**Pull request by pepone**

### CODE REVIEW

File: cpp/src/IceGrid/ServerAdapterI.h [hunk 1]
```diff
@@ -19,37 +19,37 @@ namespace IceGrid
...
             const std::string&,
-            const AdapterPrxPtr&,
+            AdapterPrx,
             const std::string&,
             bool);
         ~ServerAdapterI() override;
 
         void activateAsync(
-            std::function<void(const Ice::ObjectPrxPtr&)>, // TODO: pass by value!
+            std::function<void(const std::optional<Ice::ObjectPrx>&)>, // TODO: pass by value!
```
Comment (externl) [inline cpp/src/IceGrid/ServerAdapterI.h:28]: Is this TODO still relevant?

File: cpp/src/IceGrid/SessionManager.h [hunk 2]
```diff
@@ -140,19 +137,21 @@ namespace IceGrid
...
                             {
-                                session = createSession(registry, timeout);
+                                assert(registry);
+                                session = createSession(*registry, timeout);
                             }
                             break;
                         case Disconnect:
                             assert(session);
-                            destroySession(session);
+                            destroySession(*session);
```
Comment (externl) [inline cpp/src/IceGrid/SessionManager.h:146]: How do we now session is not null here (and below)?

File: cpp/src/IceGrid/WellKnownObjectsManager.cpp [hunk 3]
```diff
@@ -121,39 +113,36 @@ WellKnownObjectsManager::initialized() const
...
 }
 
-LocatorPrxPtr
+LocatorPrx
 WellKnownObjectsManager::getLocator()
 {
-    Ice::Identity id = {"Locator", _database->getInstanceName()};
-    return Ice::uncheckedCast<LocatorPrx>(getWellKnownObjectReplicatedProxy(std::move(id), "Client"));
+    return LocatorPrx{
+        getWellKnownObjectReplicatedProxy(Ice::Identity{"Locator", _database->getInstanceName()}, "Client")};
```
Comment (externl) [inline cpp/src/IceGrid/WellKnownObjectsManager.cpp:127]: Is Ice::Identity required here?

**Review by externl — APPROVED** [review]
Looks good to me!

File: cpp/src/IceBox/ServiceManagerI.cpp [hunk 4]
```diff
@@ -262,14 +260,22 @@ IceBox::ServiceManagerI::stopService(string name, const Current&)
...
-IceBox::ServiceManagerI::addObserver(ServiceObserverPrxPtr observer, const Current&)
+IceBox::ServiceManagerI::addObserver(optional<ServiceObserverPrx> observer, const Current&)
 {
-    //
-    // Null observers and duplicate registrations are ignored
-    //
+    // Null observers are ignored
+    if (observer)
+    {
+        addObserver(*observer);
```
[the remark this answers is not in the document]
Comment (pepone) [reply cpp/src/IceBox/ServiceManagerI.cpp:268]: Yes we should.

File: cpp/src/IceGrid/LocatorRegistryI.cpp [hunk 5]
```diff
@@ -219,22 +215,21 @@ LocatorRegistryI::setReplicatedAdapterDirectProxyAsync(
...
+
         // Get the server from the registry and set its process proxy.
         //
-        // NOTE: We pass false to the getServer call to indicate that
-        // we don't necessarily want an up-to-date adapter proxy. This
-        // is needed for the session activation mode for cases where
-        // the server is released during the server startup.
-        //
-        ServerPrxPtr server;
+        // NOTE: We pass false to the getServer call to indicate that we don't necessarily want an up-to-date adapter
```
[the remark this answers is not in the document]
Comment (pepone) [reply cpp/src/IceGrid/LocatorRegistryI.cpp:229]: I think the comment refers to the the `getProxy` call in `getServer(id)->getProxy(false);`

File: cpp/src/IceGrid/RegistryI.cpp [hunk 6]
```diff
@@ -394,16 +399,15 @@ RegistryI::startImpl()
...
-        Identity id;
-        id.category = _instanceName;
-        id.name = (_initFromReplica == "Master") ? "Registry" : "Registry-" + _initFromReplica;
+        Ice::Identity id{(_initFromReplica == "Master") ? "Registry" : "Registry-" + _initFromReplica, _instanceName};
 
-        ObjectPrxPtr proxy;
+        optional<ObjectPrx> proxy;
         try
         {
             proxy = _database->getObjectProxy(id);
```
[the remark this answers is not in the document]
Comment (pepone) [reply cpp/src/IceGrid/RegistryI.cpp:407]: I don't think so. This way we know that we are getting non null proxies from the database.

File: cpp/src/IceGrid/ServerAdapterI.h (hunk 1 above)
Comment (pepone) [reply cpp/src/IceGrid/ServerAdapterI.h:28]: Yes, the callback is still taking a reference as before.

File: cpp/src/IceGrid/SessionManager.h (hunk 2 above)
Comment (pepone) [reply cpp/src/IceGrid/SessionManager.h:146]: The previous code already has an assert for this.

File: cpp/src/IceGrid/WellKnownObjectsManager.cpp (hunk 3 above)
Comment (pepone) [reply cpp/src/IceGrid/WellKnownObjectsManager.cpp:127]: Not sure, but IMO it is more readable.

File: cpp/src/IceBox/ServiceManagerI.cpp [hunk 7]
```diff
@@ -262,14 +260,22 @@ IceBox::ServiceManagerI::stopService(string name, const Current&)
...
-IceBox::ServiceManagerI::addObserver(ServiceObserverPrxPtr observer, const Current&)
+IceBox::ServiceManagerI::addObserver(optional<ServiceObserverPrx> observer, const Current&)
 {
-    //
-    // Null observers and duplicate registrations are ignored
-    //
+    // Null observers are ignored
+    if (observer)
+    {
+        addObserver(*std::move(observer));
```
Comment (bernardnormier) [inline cpp/src/IceBox/ServiceManagerI.cpp:268]: Uses optional's
```
constexpr T&& operator*() && noexcept;
```

File: cpp/src/IceGrid/AdminI.cpp [hunk 8]
```diff
@@ -129,7 +129,7 @@ namespace
 
     private:
         string _id;
-        ServerPrxPtr _proxy;
+        optional<ServerPrx> _proxy;
```
Comment (bernardnormier) [inline cpp/src/IceGrid/AdminI.cpp:132]: I don't see how _proxy can be null. It's just convenience for the constructor?

File: slice/Ice/Locator.ice [hunk 9]
```diff
@@ -87,7 +87,7 @@ interface LocatorRegistry
     /// Set the adapter endpoints with the locator registry.
     /// @param id The adapter id.
     /// @param proxy The adapter proxy (a dummy direct proxy created by the adapter). The direct proxy contains the
-    /// adapter endpoints.
+    /// adapter endpoints. The proxy is never null.
```
Comment (bernardnormier) [inline slice/Ice/Locator.ice:90]: This is bogus. We pass a null proxy during object adapter deactivation.

File: slice/Ice/Locator.ice [hunk 10]
```diff
@@ -97,19 +97,19 @@ interface LocatorRegistry
     /// Set the adapter endpoints with the locator registry.
     /// @param adapterId The adapter id.
     /// @param replicaGroupId The replica group id.
-    /// @param p The adapter proxy (a dummy direct proxy created by the adapter). The direct proxy contains the adapter
-    /// endpoints.
+    /// @param proxy The adapter proxy (a dummy direct proxy created by the adapter). The direct proxy contains the adapter
```
Comment (bernardnormier) [inline slice/Ice/Locator.ice:100]: It can be null, see above.

File: slice/Ice/Locator.ice [hunk 11]
```diff
@@ -97,19 +97,19 @@ interface LocatorRegistry
...
     /// @throws InvalidReplicaGroupIdException Raised if the given replica group doesn't match the one registered with
     /// the locator registry for this object adapter.
-    ["amd"] idempotent void setReplicatedAdapterDirectProxy(string adapterId, string replicaGroupId, Object* p)
+    ["amd"] idempotent void setReplicatedAdapterDirectProxy(string adapterId, string replicaGroupId, Object* proxy)
         throws AdapterNotFoundException, AdapterAlreadyActiveException, InvalidReplicaGroupIdException;
 
     /// Set the process proxy for a server.
     /// @param id The server id.
-    /// @param proxy The process proxy.
+    /// @param proxy The process proxy. The proxy is never null.
```
Comment (bernardnormier) [inline slice/Ice/Locator.ice:112]: I believe this is correct - this proxy is never null.

File: slice/IceGrid/Admin.ice [hunk 12]
```diff
@@ -75,7 +75,7 @@ struct AdapterInfo
     /// The id of the adapter.
     string id;
 
-    /// A dummy direct proxy that contains the adapter endpoints.
+    /// A dummy direct proxy that contains the adapter endpoints. This proxy is never null.
```
Comment (bernardnormier) [inline slice/IceGrid/Admin.ice:78]: I don't think we can make this statement for struct fields in general. The application can always put a null proxy in a struct field.

Then, if you look at the Database implementation, this field is occasionally null.

Such information would be more valuable on an operation that accepts or returns an AdapterInfo or AdapterInfoSeq. Like (if accurate) "the proxy field of the returned adapter info is never null".

**Review by bernardnormier — APPROVED** [review]
Looks good, but the Slice file doc-comments need some fixes.

File: cpp/src/IceGrid/AdminI.cpp (hunk 8 above)
Comment (pepone) [reply cpp/src/IceGrid/AdminI.cpp:132]: We can fix this by using a helper function, and moving the `SynchronizationException` handling to the helper method.

Comment (pepone) [reply cpp/src/IceGrid/AdminI.cpp:132]: I will try to fix in a follow up PR.

**Pull request merged**